### PR TITLE
Add "last generated"

### PR DESCRIPTION
**Pull request by tioirawan**
This will add "last generated n ago" from scraped data

Closes https://github.com/coala/gci-leaders/issues/1

### CODE REVIEW

File: generate.js [hunk 1]
```diff
@@ -1,6 +1,6 @@
 const fs = require('fs')
 const Mustache = require('mustache')
-const orgs = require('./out/data.json')
+const {orgs, timestamp} = require('./out/data.json')
```
Comment (jayvdb) [inline generate.js]: Can you load the data files modified date?

Also the modified date of the original json blob on the GCI server should be considered, if it gives appropriate headers.

**Review by jayvdb — CHANGES_REQUESTED** [review]

File: generate.js (hunk 1 above)
Comment (blazeu) [reply generate.js]: Yeah, it should get the modified date of the json file.

Use [`fs.stat`](https://nodejs.org/api/fs.html#fs_fs_statsync_path)

File: scrap.js [hunk 2]
```diff
@@ -30,12 +30,13 @@ async function fetchOrgsWithLeaders() {
...
   
   // sort data by completed_task_instance_count
-  data.sort((a, b) => 
+  orgs.sort((a, b) => 
     b.completed_task_instance_count - a.completed_task_instance_count
   )
 
-  fs.writeFileSync('out/data.json', JSON.stringify(data))
+  fs.writeFileSync('out/data.json', JSON.stringify({orgs, timestamp}))
 })()
```
Comment (blazeu) [inline scrap.js]: This is not needed if you rely on the generated file's modified date.

**Review by blazeu — CHANGES_REQUESTED** [review]

File: scrap.js [hunk 3]
```diff
@@ -30,7 +30,7 @@ async function fetchOrgsWithLeaders() {
 }
 
 ;(async () => {
-  const data = await fetchOrgsWithLeaders()
+  const data= await fetchOrgsWithLeaders()
```
Comment (blazeu) [inline scrap.js]: Do not mess with whitespace in an unrelated file.

File: generate.js [hunk 4]
```diff
@@ -60,8 +63,45 @@ const template = `
...
+        }
+        
+        else if (elapsed < msPerYear) {
+          return 'approximately ' + Math.round(elapsed/msPerMonth) + ' months ago';   
+        }
+        
+        else {
+          return 'approximately ' + Math.round(elapsed/msPerYear ) + ' years ago';   
+        }
+      }
```
Comment (blazeu) [inline generate.js]: Just a heads-up, this repo will use prettier.io code formatter via ESLintBear.

Ideally you should start formatting it now (using `prettier` directly)

**Review by blazeu — CHANGES_REQUESTED** [review]

File: generate.js (hunk 4 above)
Comment (Indmind) [reply generate.js]: do I need to prettify with this https://prettier.io/playground/ ?

Comment (blazeu) [reply generate.js]: Sure that'll work for now.

Comment (blazeu) [reply generate.js]: Don't forget to set it without semicolon though :)

Comment (blazeu) [reply generate.js]: And with trailing comma

File: generate.js [hunk 5]
```diff
@@ -1,6 +1,7 @@
 const fs = require('fs')
 const Mustache = require('mustache')
 const orgs = require('./out/data.json')
+const timestamp =  new Date(fs.statSync('./out/data.json').mtime).getTime()
```
Comment (andrewda) [inline generate.js]: Remove one of the spaces after the equals sign.

We should get coala working ASAP

**Review by andrewda — CHANGES_REQUESTED** [review]

**Comment by jayvdb** [comment]
Did you investigate the GCI server timestamps for the raw data?  as mentioned [here](
https://github.com/coala/gci-leaders/pull/19#discussion_r154985995) ?  Are the timestamps stable, indicating that the server is updating that file only periodically...?

File: generate.js [hunk 6]
```diff
@@ -60,8 +63,35 @@ const template = `
...
+          return Math.round(elapsed / msPerHour) + " hours ago"
+        } else if (elapsed < msPerMonth) {
+          return "approximately " + Math.round(elapsed / msPerDay) + " days ago"
+        } else if (elapsed < msPerYear) {
+          return "approximately " + Math.round(elapsed / msPerMonth) + " months ago"
+        } else {
+          return "approximately " + Math.round(elapsed / msPerYear) + " years ago"
+        }
+      }      
+    </script>
```
Comment (blazeu) [inline generate.js]: Use single quote as well, sorry.

**Comment by jayvdb** [comment]
unack https://github.com/coala/gci-leaders/pull/19/commits/64ccdf496cafc8bf21375461cabb4159860bbc74

File: generate.js (hunk 6 above)
Comment (blazeu) [inline generate.js]: [coala](https://github.com/coala/gci-leaders/pull/20) is going to be used soon, can you move the js to it's own file, preferably inside `static` folder.

**Review by blazeu — CHANGES_REQUESTED** [review]
Please use .js file, and name it something generic like `app.js`

**Comment by tioirawan** [comment]
how about css? should I move it too?

**Comment by blazeu** [comment]
No, that should go in another task 😄 

By the way, don't forget to rebase 👌

**Comment by blazeu** [comment]
_Protip_: if you want to modify previous commit, you can use `git commit --amend` so it doesn't produce a new commit.

**Comment by tioirawan** [comment]
thanks, I try to rebase and got conflict

File: generate.js [hunk 7]
```diff
@@ -43,6 +46,8 @@ const template = `
   </head>
   <body>
     <h1>GCI Current Leaders</h1>
+    <small id="last-generated"></small>
```
Comment (blazeu) [inline generate.js]: Please put the UTC date as well, like in http://community.coala.io/
Feel free to take a look at it's source code in coala/community.

**Review by blazeu — CHANGES_REQUESTED** [review]

File: static/js/app.js [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+  } else if (elapsed < msPerDay) {
+    return Math.round(elapsed / msPerHour) + ' hours ago'
+  } else if (elapsed < msPerMonth) {
+    return 'approximately ' + Math.round(elapsed / msPerDay) + ' days ago'
+  } else if (elapsed < msPerYear) {
+    return 'approximately ' + Math.round(elapsed / msPerMonth) + ' months ago'
+  } else {
+    return 'approximately ' + Math.round(elapsed / msPerYear) + ' years ago'
+  }
+}
```
Comment (andrewda) [inline static/js/app.js:28]: We should probably be linting this in the `.coafile`. However the rules might differ from the other JavaScript tests because it's browser and not Node.js.

Comment (blazeu) [reply static/js/app.js:28]: Now that you talk about it, I forgot to told you to put real linting rules :P i.e. `eslint: recommended`

File: generate.js [hunk 9]
```diff
@@ -46,8 +48,10 @@ const template = `
...
+    <hr>
     <i>
-      The leading participants for each organization are listed alphabetically
+      The leading participants for each organization are listed alphabetically
```
Comment (jayvdb) [inline generate.js]: Line contains following spacing inconsistencies:
- Trailing whitespaces.

*Origin: SpaceConsistencyBear, Section: `all.whitespace`.*

The issue can be fixed by applying the following patch:

```diff
--- a/tmp/tmpsmvkva_5/generate.js
+++ b/tmp/tmpsmvkva_5/generate.js
@@ -51,7 +51,7 @@
     <small id="last-generated"></small>
     <hr>
     <i>
-      The leading participants for each organization are listed alphabetically 
+      The leading participants for each organization are listed alphabetically
       according to their "display name"
     </i>
     <div class="orgs">
```

File: generate.js [hunk 10]
```diff
@@ -78,8 +82,14 @@ const template = `
...
+    <script>
+      ${script}
+    </script>
   </body>
 </html>
 `
 
-fs.writeFileSync('out/index.html', Mustache.render(template, { orgs }))
+fs.writeFileSync(
+  'out/index.html',
```
Comment (jayvdb) [inline generate.js]: Line contains following spacing inconsistencies:
- Trailing whitespaces.

*Origin: SpaceConsistencyBear, Section: `all.whitespace`.*

The issue can be fixed by applying the following patch:

```diff
--- a/tmp/tmpsmvkva_5/generate.js
+++ b/tmp/tmpsmvkva_5/generate.js
@@ -90,6 +90,6 @@
 `
 
 fs.writeFileSync(
-  'out/index.html', 
+  'out/index.html',
   Mustache.render(template, { orgs, timestamp })
 )
```

File: generate.js [hunk 11]
```diff
@@ -78,8 +82,14 @@ const template = `
         </div>
       {{/orgs}}
     </div>
+    <script>
+      ${script}
```
Comment (blazeu) [inline generate.js]: `<script src="static/js/app.js"></script>`

Comment (tioirawan) [reply generate.js]: the mustache template doesn't work

File: generate.js [hunk 12]
```diff
@@ -46,6 +48,8 @@ const template = `
   </head>
   <body>
     <h1>GCI Current Leaders</h1>
+    <small id="last-generated"></small>
```
Comment (blazeu) [inline generate.js]: You can use `data` attribute here i.e. `data-time` and then get that value in js with [`.dataset`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/dataset)

File: generate.js (hunk 7 above)
Comment (blazeu) [reply generate.js]: @Indmind ^ Please do that first

File: generate.js (hunk 12 above)
Comment (tioirawan) [reply generate.js]: where should I put the static folder?

Comment (blazeu) [reply generate.js]: Your file placement is already correct

Comment (tioirawan) [reply generate.js]: it doesn't work, if i move static folder to out folder, it works

Comment (blazeu) [reply generate.js]: Hmm I see.
https://github.com/coala/gci-leaders/pull/31 introduced that behaviour of copying static to out folder.
It'll be merged soon.

You can do/test the other stuff while waiting.

**Comment by blazeu** [comment]
^ Now that's merged, you can now put js file in static folder and it'll be copied.

File: static/js/app.js [hunk 13]
```diff
@@ -0,0 +1,28 @@
+const lastGenerated = document.getElementById('last-generated')
+const timestamp = new Date(lastGenerated.dataset.time)
+
+lastGenerated.textContent = `Last updated: ${timestamp.toDateString()} (${timeDifference(new Date().getTime(), timestamp.getTime())})`
```
Comment (blazeu) [inline static/js/app.js]: Should include the time as well.
use `date.toUTCString()`

File: templates/main.html [hunk 14]
```diff
@@ -8,6 +8,8 @@
   </head>
   <body>
     <h1>GCI Current Leaders</h1>
+    <small id="last-generated" data-time="{{timestamp}}"></small>
```
Comment (blazeu) [inline templates/main.html]: You can put the absolute datetime (.. .. 2017) at build time here,
but still set the relative datetime (.. ago) with client side js.

Like in [coala/community](https://github.com/coala/community/blob/f900af1d42c9c304739953c624024216737637b2/index.html#L14)

It's also nice to have the absolute datetime inside [`<time>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/time).
But not mandatory, up to you.

**Review by blazeu — CHANGES_REQUESTED** [review]

**Comment by blazeu** [comment]
Please use the new `package-lock.json` @ https://github.com/coala/gci-leaders/commit/5d79b88d48ffdf3f27ac399e9b1a67eba69c564a and do not modify it.

Sorry for the interruption.

**Review by jayvdb — APPROVED** [review]

**Comment by andrewda** [comment]
If #37 is merged shortly, it'll still be good to add `static/js/*.js` to the linted JavaScript files.

**Comment by jayvdb** [comment]
(Im just clearing my blocker; dont have time to re-review; leaving that to others)

File: generate.js [hunk 15]
```diff
@@ -11,5 +13,5 @@ ncp('static', 'out/static', err => {
 
 fs.writeFileSync(
   'out/index.html',
-  Mustache.render(fs.readFileSync('templates/main.html').toString(), { orgs })
+  Mustache.render(fs.readFileSync('templates/main.html').toString(), { orgs, datetime })
```
Comment (blazeu) [inline generate.js]: Format this with prettier.
It's blocking the CI build.

**Review by blazeu — CHANGES_REQUESTED** [review]

**Comment by blazeu** [comment]
Nice :tada: 

Now for the final request,

do rebase first,
then add `static/js/*.js` to the first `files` key in `.coafile`,
and your .js file is now linted properly in our CI :+1:

**Comment by tioirawan** [comment]
like this?
screenshot from 2017-12-06 16-05-14

**Comment by blazeu** [comment]
Don't remove the existing `*.js` ;)
We still need those to lint files in the root e.g. `scrap.js` and `generate.js`

**Comment by Indmind** [comment]
yeah, I just notice that

**Review by blazeu — APPROVED** [review]

**Comment by blazeu** [comment]
ack https://github.com/coala/gci-leaders/pull/19/commits/c8615f314c5e60e4e3c225c348af247863e19222

**Comment by blazeu** [comment]
Ah, can you change your commit message?
You said it was "last generated" but its actually "Last updated".

You can submit your GCI task now, though ;)

Also, you might be in the leaderboard after this task, you've done enough task to be worth in there, I think.

**Comment by blazeu** [comment]
reack https://github.com/coala/gci-leaders/pull/19/commits/ad8d15a6d0e033809b151516258dfe1af1ecd53f

**Comment by tioirawan** [comment]
Done, thanks :)
